modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,454 @@
+     * @param inpBuf Input byte array.
+     * @param inpOff Start offset in the input array.
+     * @param inpLen Number of bytes to write.
+     */
+    void write(int pos, byte[] inpBuf, int inpOff, int inpLen) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < inpLen)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        Objects.checkFromIndexSize(inpOff, inpLen, inpBuf.length);
+
+        int newLen = Math.max(pos + inpLen, len);
+
+        ensureCapacity(newLen);
+
+        U.arrayCopy(inpBuf, inpOff, arr, pos, inpLen);
+
+        len = newLen;
+    }
+
+    /**
+     * Read a byte from this buffer from specified position {@code pos}.
+     *
+     * @param pos Position.
+     * @return Byte read from the Blob. -1 if end of data reached.
+     */
+    int read(int pos) {
+        if (pos >= len)
+            return -1;
+
+        return arr[pos + off] & 0xff;
+    }
+
+    /**
+     * Write a byte to this buffer to specified position {@code pos}.
+     *
+     * <p>The byte to be written is the eight low-order bits of the
+     * argument {@code b}. The 24 high-order bits of {@code b}b are ignored.
+     *
+     * @param pos Pointer to a position.
+     * @param b Byte to write.
+     */
+    void write(int pos, int b) throws IOException {
+        if (MAX_ARRAY_SIZE - pos < 1)
+            throw new IOException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes.");
+
+        int newLen = Math.max(pos + 1, len);
+
+        ensureCapacity(newLen);
+
+        arr[pos] = (byte)b;
+
+        len = newLen;
+    }
+
+    /**
+     * Ensure capacity.
+     *
+     * @param newLen The new data length the buffer should be able to hold.
+     */
+    private void ensureCapacity(int newLen) {
+        if (newLen - arr.length > 0 || isReadOnly)
+            reallocate(capacity(arr.length, newLen));
+    }
+
+    /**
+     * Calculate new capacity.
+     *
+     * @param cap Current capacity.
+     * @param reqCap Required new capacity.
+     * @return New capacity.
+     */
+    protected static int capacity(int cap, int reqCap) {
+        if (reqCap <= MIN_CAP)
+            return MIN_CAP;
+
+        int resCap = Math.max(cap, MIN_CAP);
+
+        while (resCap < reqCap) {
+            resCap <<= 1;
+
+            if (resCap < 0)
+                return MAX_ARRAY_SIZE;
+        }
+
+        return resCap;
+    }
+
+    /**
+     * Allocate the new underlining array and copy data.
+     *
+     * @param newCapacity New capacity.
+     */
+    private void reallocate(int newCapacity) {
+        byte[] newBuf = new byte[newCapacity];
+
+        U.arrayCopy(arr, off, newBuf, 0, len);
+
+        arr = newBuf;
+        off = 0;
+        isReadOnly = false;
+    }
+
+    /**
+     * Input stream to read data from buffer.
+     */
+    private class BufferInputStream extends InputStream {
+        /** Stream starting position. */
+        private final int start;
```

Comment:
We can eliminate this.

`pos` means current position in buffer.
`limit` - max position in buffer.